wlroots/wlr_types/layer_shell_v1.py

```diff
@@ -0,0 +1,213 @@
+
+    @property
+    def desired_width(self) -> int:
+        return self._ptr.desired_width
+
+    @property
+    def desired_height(self) -> int:
+        return self._ptr.desired_height
+
+    @property
+    def actual_width(self) -> int:
+        return self._ptr.actual_width
+
+    @property
+    def actual_height(self) -> int:
+        return self._ptr.actual_height
+
+    @property
+    def layer(self) -> LayerShellV1Layer:
+        return LayerShellV1Layer(self._ptr.layer)
+
+
+class LayerSurfaceV1(Ptr):
+    def __init__(self, ptr):
+        self._ptr = ffi.cast("struct wlr_layer_surface_v1 *", ptr)
+
+        self.destroy_event = Signal(ptr=ffi.addressof(self._ptr.events.destroy))
+        self.map_event = Signal(ptr=ffi.addressof(self._ptr.events.map))
+        self.unmap_event = Signal(ptr=ffi.addressof(self._ptr.events.unmap))
+        self.new_popup_event = Signal(ptr=ffi.addressof(self._ptr.events.new_popup))
+
+    @property
+    def surface(self) -> Surface:
+        surface_ptr = self._ptr.surface
+        _weakkeydict[surface_ptr] = self._ptr
+        return Surface(surface_ptr)
+
+    @property
+    def output(self) -> Optional[Output]:
+        output_ptr = self._ptr.output
+        if output_ptr == ffi.NULL:
+            return None
+        _weakkeydict[output_ptr] = self._ptr
+        return Output(output_ptr)
+
+    @output.setter
+    def output(self, output: Output):
+        self._ptr.output = output._ptr
+
+    @property
+    def added(self) -> bool:
+        return self._ptr.added
+
+    @property
+    def configured(self) -> bool:
+        return self._ptr.configured
+
+    @property
+    def mapped(self) -> bool:
+        return self._ptr.mapped
+
+    @property
+    def closed(self) -> bool:
+        return self._ptr.closed
+
+    @property
+    def client_pending(self) -> LayerSurfaceV1State:
+        state_ptr = self._ptr.client_pending
+        _weakkeydict[state_ptr] = self._ptr
+        return LayerSurfaceV1State(state_ptr)
+
+    @property
+    def server_pending(self) -> LayerSurfaceV1State:
+        state_ptr = self._ptr.server_pending
+        _weakkeydict[state_ptr] = self._ptr
+        return LayerSurfaceV1State(state_ptr)
+
+    @property
+    def current(self) -> LayerSurfaceV1State:
+        state_ptr = self._ptr.current
+        _weakkeydict[state_ptr] = self._ptr
+        return LayerSurfaceV1State(state_ptr)
+
+    def configure(self, width: int, height: int) -> None:
+        """
+        Notifies the layer surface to configure itself with this width/height. The
+        layer_surface will signal its map event when the surface is ready to assume
+        this size.
+        """
+        lib.wlr_layer_surface_v1_configure(self._ptr, width, height)
+
+    def close(self) -> None:
+        lib.wlr_layer_surface_v1_close(self._ptr)
+
+    @classmethod
+    def from_wlr_surface(cls, surface: Surface):
+        surface_ptr = lib.wlr_layer_surface_v1_from_wlr_surface(surface._ptr)
```

Comment:
I think it would be good to first check `lib.wlr_surface_is_layer_surface` for this surface and throw an exception if it is not. I think if the `wlr_layer_surface_v1_from_wlr_surface` fails, it does so by aborting.

Reply:
Yep sounds good

Reply:
Well, it does make sense, but would this not be the responsibility of the using code to ensure? They may have already confirmed by some other means so having it always check might be an extra function call when not always necessary.

Reply:
That is a good point. I think the only possible problem there is that if this fails, then we'll get some sort of fault from in the library that Python doesn't know how to handle, rather than some sort of Python exception. For now, this is fine, but it's something to consider if we get bit by this.